pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,419 @@
+	for gc.attemptToDeleteWorker() {
+	}
+}
+
+func (gc *GarbageCollector) attemptToDeleteWorker() bool {
+	item, quit := gc.attemptToDelete.Get()
+	if quit {
+		return false
+	}
+	defer gc.attemptToDelete.Done(item)
+	n, ok := item.(*node)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expect *node, got %#v", item))
+		return true
+	}
+	err := gc.attempToDeleteItem(n)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing item %#v: %v", n, err))
+		// retry if garbage collection of an object failed.
+		gc.attemptToDelete.AddRateLimited(item)
+	}
+	return true
+}
+
+func objectReferenceToMetadataOnlyObject(ref objectReference) *metaonly.MetadataOnlyObject {
+	return &metaonly.MetadataOnlyObject{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: ref.APIVersion,
+			Kind:       ref.Kind,
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Namespace: ref.Namespace,
+			UID:       ref.UID,
+			Name:      ref.Name,
+		},
+	}
+}
+
+// classify the latestReferences to three categories:
+// solid: the owner exists, and is not "waitingForDependentsDeletion"
+// dangling: the owner does not exist
+// waitingForDependentsDeletion: the owner exists, its deletionTimestamp is non-nil, and it has
+// FinalizerDeletingDependents
+// This function communicates with the server.
+func (gc *GarbageCollector) classifyReferences(item *node, latestReferences []metav1.OwnerReference) (
+	solid, dangling, waitingForDependentsDeletion []metav1.OwnerReference, err error) {
+	for _, reference := range latestReferences {
+		if gc.absentOwnerCache.Has(reference.UID) {
+			glog.V(5).Infof("according to the absentOwnerCache, object %s's owner %s/%s, %s does not exist", item.identity.UID, reference.APIVersion, reference.Kind, reference.Name)
+			dangling = append(dangling, reference)
+			continue
+		}
+		// TODO: we need to verify the reference resource is supported by the
+		// system. If it's not a valid resource, the garbage collector should i)
+		// ignore the reference when decide if the object should be deleted, and
+		// ii) should update the object to remove such references. This is to
+		// prevent objects having references to an old resource from being
+		// deleted during a cluster upgrade.
+		fqKind := schema.FromAPIVersionAndKind(reference.APIVersion, reference.Kind)
+		client, err := gc.clientPool.ClientForGroupVersionKind(fqKind)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		resource, err := gc.apiResource(reference.APIVersion, reference.Kind, len(item.identity.Namespace) != 0)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		// TODO: It's only necessary to talk to the API server if the owner node
+		// is a "virtual" node. The local graph could lag behind the real
+		// status, but in practice, the difference is small.
+		owner, err := client.Resource(resource, item.identity.Namespace).Get(reference.Name)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return nil, nil, nil, err
+			}
+			gc.absentOwnerCache.Add(reference.UID)
```

Comment:
I think this would be a little easier to read since its more positive logic

```go
switch{
case isnotfound(err):
    // do this logic
case err != nil:
    return nil, nil, nil, err
}
```